core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala

```diff
@@ -1176,6 +1177,31 @@ class RDDSuite extends SparkFunSuite with SharedSparkContext {
     }.collect()
   }
 
+  test("SPARK-27666: Do not release lock while TaskContext already completed") {
+    val rdd = sc.parallelize(Range(0, 10), 1).cache()
+    val tid = sc.longAccumulator("threadId")
+    // validate cache
+    rdd.collect()
+    rdd.mapPartitions { iter =>
+      val t = new Thread(() => {
+        while (iter.hasNext) {
+          iter.next()
+          Thread.sleep(100)
+        }
+      })
+      t.setDaemon(false)
+      t.start()
+      tid.add(t.getId)
+      Iterator(0)
+    }.collect()
+    val tmx = ManagementFactory.getThreadMXBean
+    var t = tmx.getThreadInfo(tid.value)
+    // getThreadInfo() will return null after child thread `t` died
+    while (t != null && t.getThreadState != Thread.State.TERMINATED) {
+      t = tmx.getThreadInfo(tid.value)
+    }
+  }
```

Comment:
Can we use eventually with timeout instead of a while loop?

```scala
eventually(timeout(10.seconds))  {
  val t = tmx.getThreadInfo(tid.value)
  assert(t == null || t.getThreadState == Thread.State.TERMINATED)
}
```

Reply:
good idea!